In [49]:
import pandas as pd
reviews = pd.read_csv('../data/wine/winemag-data_first1000.csv', index_col=0)
reviews

country                                        description  \
0           US  This is quite soft and pretty, a charming wine...   
1       France  A crisp, fresh Riesling that manages to be lea...   
2           US  A medium bouquet of apple, lemon and a touch o...   
3    Argentina  Aromas of blackberry, balsam wood, lemony oak ...   
4        Italy  Aromas of wild flower, honey and peach charact...   
..         ...                                                ...   
995         US  Soft, slightly sweet and simple, this Syrah ha...   
996         US  Overripe, with blackberry jam fruit veering in...   
997     France  Big, full-bodied wine, the great tannins domin...   
998         US  Hogue's lowest-priced Chard is reliably clean ...   
999  Australia  Vanilla, plum and a tangy, citrusy note domina...   

                            designation  points  price          province  \
0                       Vineyard Select      87   18.0            Oregon   
1                             Le Dragon      86   34.0            Alsace   
2                             Grand Cru      85   15.0          New York   
3                               Reserva      90   24.0  Mendoza Province   
4    Villa Luigia Millesimato Extra Dry      87   20.0            Veneto   
..                                  ...     ...    ...               ...   
995                Amber Ridge Vineyard      85   29.0        California   
996                  Dryfarmed Old Vine      84   25.0        California   
997                                 NaN      94  190.0          Burgundy   
998                                 NaN      85    9.0        Washington   
999                      Bin Select 667      83    7.0   Australia Other   

                                        region_1           region_2  \
0                             Chehalem Mountains  Willamette Valley   
1                                         Alsace                NaN   
2                      North Fork of Long Island        Long Island   
3                                        Mendoza                NaN   
4    Conegliano Valdobbiadene Prosecco Superiore                NaN   
..                                           ...                ...   
995                         Russian River Valley             Sonoma   
996                               Cienega Valley      Central Coast   
997                           Charmes-Chambertin                NaN   
998                         Columbia Valley (WA)    Columbia Valley   
999                      South Eastern Australia                NaN   

              variety                winery  
0          Pinot Noir               Gresser  
1            Riesling              Josmeyer  
2          Chardonnay               Raphael  
3              Malbec      Sur de los Andes  
4               Glera              Il Follo  
..                ...                   ...  
995             Syrah                  Novy  
996  Cabernet Pfeffer                DeRose  
997        Pinot Noir  Chanson Père et Fils  
998        Chardonnay                 Hogue  
999            Shiraz              Kalbarri  

[1000 rows x 10 columns]

## Exercise 1

Επιλέξτε τη στήλη `description` από τo Dataframe `reviews` και αντιστοιχίστε το αποτέλεσμα στη μεταβλητή `desc`.

In [30]:
desc = reviews['description']


In [31]:
#or
desc = reviews.description


desc is a pandas Series object, with an index matching the reviews DataFrame. In general, when we select a single column from a DataFrame, we'll get a Series.

## Exercise 2

Επιλέξτε την πρώτη τιμή από τη στήλη `description`, εκχωρώντας τη στη μεταβλητή `first_description`.

In [32]:
first_description = reviews.description.iloc[0]
first_description

'This is quite soft and pretty, a charming wine with appealing flavors of rose water, cherry and sandalwood. Not for keeping, but most enjoyable right now.'

Note that while this is the preferred way to obtain the entry in the DataFrame, many other options will return a valid result, such as `reviews.description.loc[0]`, `reviews.description[0]`, and more!

## Exercise 3

Επιλέξτε την πρώτη σειρά δεδομένων (την πρώτη εγγραφή) από το `Dataframe`, εκχωρώντας τη στη μεταβλητή `first_row`.

In [33]:
first_row = first_row = reviews.iloc[0]
first_row

country                                                       US
description    This is quite soft and pretty, a charming wine...
designation                                      Vineyard Select
points                                                        87
price                                                       18.0
province                                                  Oregon
region_1                                      Chehalem Mountains
region_2                                       Willamette Valley
variety                                               Pinot Noir
winery                                                   Gresser
Name: 0, dtype: object

## Exercise 4

Επιλέξτε τις πρώτες 10 τιμές από τη στήλη `description`, εκχωρώντας το αποτέλεσμα στη μεταβλητή `first_descriptions`.

Συμβουλή: μορφοποιήστε την έξοδο σας ως pandas Series.

In [34]:
first_descriptions = reviews.description.iloc[0:10]
first_descriptions

0    This is quite soft and pretty, a charming wine...
1    A crisp, fresh Riesling that manages to be lea...
2    A medium bouquet of apple, lemon and a touch o...
3    Aromas of blackberry, balsam wood, lemony oak ...
4    Aromas of wild flower, honey and peach charact...
5    Nice aromas suggest cassis and deep, wild moun...
6    A worthy followup to the 2001, it's not as vol...
7    Distinctly different from the Avery vineyard b...
8    This gets the award for the most confusing lab...
9    Unique in style and sourced—these are the olde...
Name: description, dtype: object

Note that many other options will return a valid result, such as `desc.head(10)` and `reviews.loc[:9, "description"]`.



## Exercise 5

Επιλέξτε τις εγγραφές με τις ετικέτες ευρετηρίου 1, 2, 3, 5 και 8, εκχωρώντας το αποτέλεσμα στη μεταβλητή `sample_reviews`.

In [35]:
indicies = [1, 2, 3, 5, 8]
sample_reviews = reviews.loc[indicies]

## Exercise 6


Δημιουργήστε μια μεταβλητή `df` που περιέχει τις στήλες , `country`, `province`, `region_1`, και `region_2` των εγγραφών με τις ετικέτες ευρετηρίου `0`, `1`, `10`, και `100`.  Με άλλα λόγια, δημιουργήστε το ακόλουθο DataFrame:

![](https://i.imgur.com/FUCGiKP.png)


In [39]:
indicies = [0, 1, 10, 100]
columns = ['country', 'province', 'region_1', 'region_2']
df = reviews[columns].loc[indicies]
df

country    province              region_1           region_2
0        US      Oregon    Chehalem Mountains  Willamette Valley
1    France      Alsace                Alsace                NaN
10       US  Washington  Columbia Valley (WA)    Columbia Valley
100      US  California         Amador County   Sierra Foothills

## Exercise 7

Δημιουργήστε ένα DataFrame `df` που περιέχει τις στήλες `country` και `variety` των πρώτων 100 εγγραφών.

Συμβουλή: μπορείτε να χρησιμοποιήσετε `loc` ή `iloc`. Όταν εργάζεστε για την απάντηση σε αυτήν την ερώτηση και σε πολλές από αυτές που ακολουθούν, κρατήστε το παρακάτω:

> Το «iloc» χρησιμοποιεί το σχήμα ευρετηρίου Python stdlib, όπου περιλαμβάνεται το πρώτο στοιχείο της περιοχής και το τελευταίο εξαιρείται. Tο `loc` εμανίζει και τα included.

> Αυτό προκαλεί ιδιαίτερη σύγχυση όταν το DataFrame είναι μια απλή αριθμητική λίστα, π.χ. `0,...,1000`. Σε αυτήν την περίπτωση το `df.iloc[0:1000]` θα επιστρέψει 1000 καταχωρήσεις, ενώ το `df.loc[0:1000]` θα επιστρέψει 1001 !!! Για να λάβετε 1000 στοιχεία χρησιμοποιώντας το "loc", θα χρειαστεί να πάτε ένα χαμηλότερα και να ζητήσετε "df.iloc[0:999]".


In [40]:
cols = ['country', 'variety']
df = reviews[cols].loc[:99]
df

country                variety
0          US             Pinot Noir
1      France               Riesling
2          US             Chardonnay
3   Argentina                 Malbec
4       Italy                  Glera
..        ...                    ...
95     France                   Rosé
96   Portugal       Touriga Nacional
97         US     Cabernet Sauvignon
98     France        Sauvignon Blanc
99  Australia  Rhône-style Red Blend

[100 rows x 2 columns]

In [53]:
#or
cols_idx = [0, 8]
df_alternative = reviews.iloc[:100, cols_idx]
df_alternative

country                variety
0          US             Pinot Noir
1      France               Riesling
2          US             Chardonnay
3   Argentina                 Malbec
4       Italy                  Glera
..        ...                    ...
95     France                   Rosé
96   Portugal       Touriga Nacional
97         US     Cabernet Sauvignon
98     France        Sauvignon Blanc
99  Australia  Rhône-style Red Blend

[100 rows x 2 columns]

## Exercise 8

Δημιουργήστε ένα DataFrame `italian_wines` που περιέχει κριτικές κρασιών που παρασκευάζονται στην Ιταλία.

In [54]:
italian_wines = reviews[reviews.country == 'Italy']
italian_wines

country                                        description  \
4     Italy  Aromas of wild flower, honey and peach charact...   
13    Italy  Nicely linear and compact, this wine has many ...   
21    Italy  This gorgeous Pinot Grigio displays the fresh ...   
25    Italy  Gianni Brunelli's elegant Brunello opens with ...   
29    Italy  The evident oak element in this hearty Brunell...   
..      ...                                                ...   
952   Italy  Aromas of graphite, black currant, plum, cedar...   
985   Italy  This focused Rosso di Montefalco (Sangiovese, ...   
990   Italy  Aromas of chocolate, mocha and raisin lead the...   
992   Italy  Simple, bright and genuine, this Cabernet Sauv...   
993   Italy  This Nero d'Avola from the Agrigento area of s...   

                            designation  points  price            province  \
4    Villa Luigia Millesimato Extra Dry      87   20.0              Veneto   
13                               Litina      89    NaN            Piedmont   
21                                  NaN      90   15.0  Northeastern Italy   
25                                  NaN      88   75.0             Tuscany   
29                               Franci      90   65.0             Tuscany   
..                                  ...     ...    ...                 ...   
952                            Lodovico      95  419.0             Tuscany   
985                                 NaN      90   35.0       Central Italy   
990                                 NaN      86   45.0            Piedmont   
992                          Collection      85   10.0  Northeastern Italy   
993                                 NaN      88   14.0   Sicily & Sardinia   

                                        region_1 region_2             variety  \
4    Conegliano Valdobbiadene Prosecco Superiore      NaN               Glera   
13                      Barbera d'Asti Superiore      NaN             Barbera   
21                                    Alto Adige      NaN        Pinot Grigio   
25                        Brunello di Montalcino      NaN   Sangiovese Grosso   
29                        Brunello di Montalcino      NaN   Sangiovese Grosso   
..                                           ...      ...                 ...   
952                                      Toscana      NaN           Red Blend   
985                             Montefalco Rosso      NaN           Red Blend   
990                                       Barolo      NaN            Nebbiolo   
992                                     Trentino      NaN  Cabernet Sauvignon   
993                                      Sicilia      NaN        Nero d'Avola   

                       winery  
4                    Il Follo  
13            Cascina Castlet  
21   Kellerei Kaltern Caldaro  
25            Gianni Brunelli  
29                      Tassi  
..                        ...  
952         Tenuta di Biserno  
985                Lungarotti  
990           460 Casina Bric  
992                     Cavit  
993              D'Alessandro  

[154 rows x 10 columns]

## Exercise 9

Δημιουργήστε ένα DataFrame `top_south_europe` που περιέχει όλες τις κριτικές με τουλάχιστον 95 βαθμούς (από τους 100) για κρασιά από την Ιταλία, την Ισπανία ή την Ελλάδα .

In [60]:
top_south_europe = reviews.loc[
    (reviews.country.isin(['Italy', 'Greece', 'Spain']))
    & (reviews.points >= 90)]

top_south_europe

country                                        description  \
21    Italy  This gorgeous Pinot Grigio displays the fresh ...   
29    Italy  The evident oak element in this hearty Brunell...   
30    Italy  This 80-20 blend of Cabernet Sauvignon and Cab...   
38    Italy  San Rustico consistently makes some of the bes...   
71    Italy  Aromas of cassis, cola and a hint of lemon mak...   
..      ...                                                ...   
933   Italy  Always a fine choice among super Tuscans, Tass...   
949   Italy  Here's a youthful but oak-driven blend that of...   
952   Italy  Aromas of graphite, black currant, plum, cedar...   
985   Italy  This focused Rosso di Montefalco (Sangiovese, ...   
994   Spain  Sweet and welcoming; a good deal among pedigre...   

             designation  points  price            province  \
21                   NaN      90   15.0  Northeastern Italy   
29                Franci      90   65.0             Tuscany   
30                   NaN      91   35.0             Tuscany   
38                   NaN      90   54.0              Veneto   
71   Don Alberto Riserva      90   40.0             Tuscany   
..                   ...     ...    ...                 ...   
933            Tassinaia      92   45.0             Tuscany   
949                  NaN      90    NaN             Tuscany   
952             Lodovico      95  419.0             Tuscany   
985                  NaN      90   35.0       Central Italy   
994                  NaN      90   16.0             Levante   

                                region_1 region_2  \
21                            Alto Adige      NaN   
29                Brunello di Montalcino      NaN   
30                    Bolgheri Superiore      NaN   
38   Amarone della Valpolicella Classico      NaN   
71                      Chianti Classico      NaN   
..                                   ...      ...   
933                              Toscana      NaN   
949                   Bolgheri Superiore      NaN   
952                              Toscana      NaN   
985                     Montefalco Rosso      NaN   
994                              Jumilla      NaN   

                           variety                    winery  
21                    Pinot Grigio  Kellerei Kaltern Caldaro  
29               Sangiovese Grosso                     Tassi  
30                       Red Blend      Ferrari Iris e Figli  
38   Corvina, Rondinella, Molinara               San Rustico  
71                      Sangiovese                Le Miccine  
..                             ...                       ...  
933                      Red Blend    Castello del Terriccio  
949                      Red Blend      Ferrari Iris & Figli  
952                      Red Blend         Tenuta di Biserno  
985                      Red Blend                Lungarotti  
994                     Monastrell                  Juan Gil  

[62 rows x 10 columns]